### DISCUSSION TITLE
fsck msdos

### DISCUSSION THREAD

**Discussion by tyama501** [jbruchon/elks#1819] [General]
I found old fsck FAT source code from BSD.
https://github.com/boundarydevices/fsck_msdos

Is there a chance to porting this?

**Comment by ghaerr** [comment]
Hello @tyama501,

Wow, nice find! It appears that `fsck_msdos` is a command that has origins in NetBSD and OpenBSD. This version appears to be very similar to one used in [Android](https://android.googlesource.com/platform/external/fsck_msdos/+/refs/heads/main), and I found what looks like a more complex version [here](https://github.com/rockcarry/fsck_msdos/tree/master) which was enhanced by the Linux Foundation.

Are you thinking we should port this to ELKS, or just run on Linux or macOS? Just to try it, I was able to compile it and run on macOS using the following changes:
```
diff --git a/fsutil.h b/fsutil.h
index 7acfdd6..9f2a92f 100644
--- a/fsutil.h
+++ b/fsutil.h
@@ -4,4 +4,7 @@
 #define pwarn printf
 #define pfatal printf
 
+typedef off_t   loff_t;
+#define lseek64 lseek
+
 #endif
```
Then compile it and test on an ELKS FAT image using:
```
$ cc -Dlint *.c -o fsck_msdos
$ ./fsck_msdos -pf hd32-fat.img
** hd32-fat.img
** Phase 1 - Read and Compare FATs
Attempting to allocate 62 KB for FAT
Attempting to allocate 62 KB for FAT
** Phase 2 - Check Cluster Chains
** Phase 3 - Checking Directories
** Phase 4 - Checking for Lost Files
326 files, 29541 free (29541 clusters)
```
I was also able to find a [OpenBSD man page](https://man.openbsd.org/fsck_msdos.8) for it.

I would say it should be pretty straightforward to port to ELKS, but this needs to be looked at further as it does seem to do lots of `malloc`s for each directory entry found. I can't guess at this point where it might run out of memory. We might also want to look at what the more complex version does that this version doesn't, although that may use up even more memory on ELKS.

I would suggest playing with this on your PC by compiling it on your OS, then testing against various "bad" FAT disk images to see if it works well fixing them. If you like its operation, I can certainly help port it, or assist you in doing so for ELKS.

Thank you!

**Reply by ghaerr** [reply]
I got the [more enhanced](https://github.com/rockcarry/fsck_msdos/tree/master) version of `fsck_msdos` to compile using the following diff. It appears to check or otherwise compare FAT fragments which the original version may not. I think this enhanced version may use a lot more memory than the first, which could be a problem for ELKS.

Nonetheless, it would be interesting to compare this version with the simpler version on the bad FAT images you have, to see the differences.
```
diff --git a/fatcache.h b/fatcache.h
index 2f01608..9050387 100755
--- a/fatcache.h
+++ b/fatcache.h
@@ -31,8 +31,8 @@
 #include "dosfs.h"
 #include "tree.h"
 #include "stddef.h"
-#include <cutils/log.h>
-#include <android/log.h>
+//#include <cutils/log.h>
+//#include <android/log.h>
 #define EMPTY_FAT   (( struct cluster_chain_descriptor*)0)
 #define EMPTY_CACHE (( struct fatcache*)0)
 #define BIT(x,n)    (((x)>>(n)) & 0x1)
@@ -41,6 +41,7 @@
 #define CLEAR_BIT(x,n)  do{ \
     x &= ~(1<<n);}while(0)
 
+#if 0
 /*
  *print information when handle cluster chain
  */
@@ -52,6 +53,12 @@
 #define fsck_warn       FSCK_SLOGW
 #define fsck_err        FSCK_SLOGE
 #define fsck_debug      FSCK_SLOGD
+#else
+#define fsck_info(...)  printf(__VA_ARGS__)
+#define fsck_warn(...)  printf(__VA_ARGS__)
+#define fsck_err(...)   printf(__VA_ARGS__)
+#define fsck_debug(...) printf(__VA_ARGS__)
+#endif
 
 #ifndef offsetof
 #define offsetof(TYPE, MEMBER) ((size_t) &((TYPE *)0)->MEMBER)
diff --git a/fragment.c b/fragment.c
index 19750b1..0b81568 100755
--- a/fragment.c
+++ b/fragment.c
@@ -26,7 +26,6 @@
 *IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE
 */
 #include "fragment.h"
-#include "malloc.h"
 #include "fatcache.h"
 static int fsck_msdos_fragment_compare(struct fragment *frag1 ,struct fragment *frag2)
 {
diff --git a/fsutil.h b/fsutil.h
index 11eed1c..5ace108 100755
--- a/fsutil.h
+++ b/fsutil.h
@@ -4,4 +4,10 @@
 #define pwarn  printf
 #define pfatal printf
 
+#include <sys/types.h>
+#include <stdlib.h>
+#include <unistd.h>
+typedef off_t   loff_t;
+#define lseek64 lseek
+
 #endif
```
Running produces:
```
./fsck_msdos -pf hd32-fat.img 
** /hd32-fat.img
sectors Per cluster :2 
boot->ClusterOffset :277 
FAT16 Fielsystem
Using cluster_chain_descriptor
Total clusters 31613 
** Phase 1 - Read and Compare FATs
Attempting to allocate 62 KB for FAT
Attempting to allocate 62 KB for FAT
Begin to compare FAT
** Phase 2 - Check Cluster Chains 
Begin to handle the cluster chain
** Phase 3 - Checking Directories
rootDir->head :0
** Phase 4 - Checking for Lost Files
Verify Filesystem information
326 files, 29541 free (29541 clusters)
free_rb_tree 
free_fragment_tree
free_fragment_tree
free_fragment_tree: rb_tree is empty 
```

**Reply by tyama501** [reply]
Thank you! I will try it.

**Comment by toncho11** [comment]
In the extended version there is a:

`#define FAT_COMPARE_MAX_KB 4096`

Maybe we can set it to 300 KBs (to be compatible with a system with 640 KB of memory) on ELKS and hopefully the rest will work OK.
I think having it on ELKS is a good idea.

**Comment by ghaerr** [comment]
> In the extended version there is a:
#define FAT_COMPARE_MAX_KB 4096

Good catch. I now see the entire FAT is `malloc`d and read into. This means that, without lots of __far modifications to source code, the max FAT size would be 64k minus heap+stack = ~52k. For a FAT16 volume, that would limit the max FAT volume size to (52k / 2 * 512 =) ~13MB. So, basically only good for floppies, if ported to run on ELKS. The tool could be useful as part of the ELKS tool set for running on the host however.

If ~300k of RAM were allocated into far memory, that would only increase the max FAT volume checkable to be 72MB, smaller than most all FAT32 volumes out there. It appears the 4096k current FAT size limit would allow for max FAT32 volume size of (4096k / 4 * 512) = 524MB. Still not very large, although the 4096k define could be easily increased on 32- or 64-bit hosts.

The following source comment was kind of funny:
```
/*
 * If the FAT > this size then skip comparing, lest we risk
 * OOMing the framework. in the future we need to just re-write
 * this whole thing and optimize for less memory
 */
#define FAT_COMPARE_MAX_KB 4096
```
Rewriting software... the story of our lives? Lol.

We'll see what @tyama501 reports as to whether it can repair his corrupted floppies... ?

**Reply by toncho11** [reply]
But 72 MBs will still be good. ELKS' HDD images are 30 MBs if I remember correctly. So it could be used for self-repair from ELKS?

**Reply by ghaerr** [reply]
The max MINIX v1 image is 65Mb, but there are lots of much larger FAT images out there. 

Out of curiosity, how big are the non-floppy FAT HDD's or USB drives you have? Any under 13M? Many under 72M? 

One of the best attributes of the FAT32 filesystem for ELKS is its ability to hold far more data than MINIX v1 can.

Looking further at the source, there are lots of additional `calloc` and `malloc` allocations, one for every directory entry as well, so the usable volume size is going to be much smaller than my initial estimate. The program would have to be painstakingly rewritten. 

We might luck out and find it work for 1.44M floppy images on ELKS without a rewrite though.

If you'd like to rewrite it, I'll help, but almost every `char *` will have to be rewritten `char __far *`, and even then the user will find it won't work with their large USB thumb drive. It appears we have yet to find a `chkdsk` that works as-is for any FAT filesystem in real mode, like DOS. Perhaps rewriting the actual DOS `chkdsk` might be a better idea (IIRC it was written in ASM though)?

**Comment by tyama501** [comment]
Thank you for the discussions.

I tried first one (boundarydevices/fsck_msdos) on FreeBSD.
Unfortunately, it didn't worked saying bad signature for fd1232-pc98.img but it worked for fd1440-pc98.img!

Initial image,
df Used 1074
![df_before_vi_1440](https://github.com/ghaerr/elks/assets/61556504/b3ac4bee-8ebe-4bd0-8a2b-c336a5258c14)

I created Bad clusters with vi. (creating file and removing it)
df Used 1077
![df_after_vi_1440](https://github.com/ghaerr/elks/assets/61556504/ac42a087-549d-402f-83fc-05e2164f81d1)

Then I run the fsck_msck on FreeBSD. Fixing FAT difference and clusters.
![fsck_msdos_vi_1440](https://github.com/ghaerr/elks/assets/61556504/150b4f7c-a58c-4b1b-b18b-96c6e0e407ab)

Check it again in ELKS.
Here it is. 
df Used 1074 again!
![df_after_fix_1440](https://github.com/ghaerr/elks/assets/61556504/a12fa8dd-912d-4485-aecb-5bbc4dad9eb6)

It would be nice if we could port this to ELKS (and supporting 1232.img)

Thank you!

**Reply by ghaerr** [reply]
Thank you for the quick testing! It sounds like the program could work unmodified on ELKS if limited to only 1440k floppies. I would think that could be a pretty straight-forward port. It would likely immediately die or crash on any volume larger than 2880k though.

It is nice that the simple version (boundarydevices/fsck_msdos) corrected the lost cluster chain from the `vi` bug.

The errors that it picks up from the differences between FAT 0 and FAT 1 will always appear - ELKS does not maintain the second FAT copy, for speed reasons (this should not be a big deal, but the error will always show and be corrected).

> if we could port this to ELKS (and supporting 1232.img)

I suspect the problem with PC-98 1232k volumes are that the sector size is 1K rather than 512, unless there are also differences in the PC-98 BPB?

If the program were ported to ELKS for floppies only, you might be able to figure out the changes for 1232k support?

**Reply by tyama501** [reply]
>The errors that it picks up from the differences between FAT 0 and FAT 1 will always appear - ELKS does not maintain the second FAT ?>copy, for speed reasons (this should not be a big deal, but the error will always show and be corrected).

Yes, I know.

Signature is not in 512Byte border. Well, may be we can bypass that.
I will try porting when I got more time!
Thank you!

**Reply by ghaerr** [reply]
> Signature is not in 512Byte border. Well, may be we can bypass that.

Got it, its at byte 1022 and 1023. Here's the failing code, did you get this error message?:
```
int
readboot(dosfs, boot)
    int dosfs;
    struct bootblock *boot;
{
    u_char block[DOSBOOTBLOCKSIZE];
    u_char fsinfo[2 * DOSBOOTBLOCKSIZE];
    u_char backup[DOSBOOTBLOCKSIZE];
    int ret = FSOK;

    if (read(dosfs, block, sizeof block) < sizeof block) {
        perror("could not read boot block");
                exit(2);
    }

    if (block[510] != 0x55 || block[511] != 0xaa) {
        pfatal("Invalid signature in boot block: %02x%02x", block[511], block[510]);
                exit(2);
    }
```

> I will try porting when I got more time!

If you think its worthwhile to have this running for 1440k floppies only on ELKS, I could take an initial pass at it, and you could modify for PC-98. I was initially thinking, replying to @toncho11 that it should work for HDDs as well, but I suppose it would be useful for just floppies, given our current `vi` bug.

**Reply by tyama501** [reply]
That's great! Thank you.
Yes, I think I saw that message.

**Comment by ghaerr** [comment]
Well, I ported fsck_msdos to 16-bit ELKS as `fsck-dos`. It took about an hour, the biggest issue was generating BSD-style `uint_t` types and dealing with overflow-to-zero operations on `int` (16-bit for ELKS) when shifting left 16 or 24 bits. 

After that, the program was too large (at 24k, only seven larger) to fit on our 1440k floppy, so rebuilt it on a 2880k floppy. The program ran out of memory trying to check that floppy size. Thinking it might work on our 1.4M floppy, programs were removed to make space on the 1440k floppy and tested. No luck, it seems I much underestimated the amount of memory required. `fsck-dos` won't run on ELKS, even on our basic boot floppy, without conversion to using __far data model, which will be lots of work.

Here's the 1440k disk showing memory requirements being exceeded:
<img width="788" alt="1440" src="https://github.com/ghaerr/elks/assets/11985637/f4397673-5931-454e-ba75-92a50b24fa0e">

Conclusion: even though fsck_msdos seemed like a small program easily portable to a 16-bit OS, it was written assuming large amounts of free memory are available, in typical 32-bit programming style. Unfortunately too much memory is required for operation on ELKS 1440k floppy without a large porting effort, which even then won't work for typically seen FAT volume sizes.

**Reply by toncho11** [reply]
Can we skip some of the Phase checks? It looks like it fails at Phase 3. Would it be still useful without Phase 3?

**Reply by ghaerr** [reply]
One would have to dig into the program to understand how it works. I would not think that removing filesystem checks would be a good idea for a filesystem checker though - this could lead users into thinking that a volume was OK after being checked by our diminished `fsck-dos`, when in fact it was corrupted.

**Comment by ghaerr** [comment]
It seems a proper CHKDSK is pretty complicated... the [MSDOS 2.0 CHKDSK](https://github.com/microsoft/MS-DOS/tree/master/v2.0/source) source, written in ASM is three files: CHKDSK.ASM (900 lines), CHKPROC.ASM (1400 lines) and CHKMES.ASM (475 lines).  I also found [FreeDOS chkdsk](https://github.com/TritonDataCenter/sdcboot/tree/master/freedos/source/chkdsk) which is huge: 7 main C source files and 10 additional directories of library routines, all included in the checker. I'm not sure how either of them work, but am imagining that both use much more than 64K data in order to function.  

If we had large model support allowing for all __far pointers and automatic far memory allocation, things would be a lot easier, but that's another even larger problem and would likely require porting the Watcom C compiler to produce ELKS executables, along with a specially-created large model version of our C library - a lot of work.

Another idea would be to create and/or run a 386-based kernel on our ancient systems (386+ only, of course). This would allow for full 32-bit protected mode, even on very old systems (but still newer than 8086 and 80286). I have been investigating such a thing if anyone is interested.

**Reply by tyama501** [reply]
Thank you for your efforts. Can I have the codes you have modified? 
I may think using far memory when I have time.

**Reply by tyama501** [reply]
BTW, I have been watching your activity on Fiwix. Very impressive!

**Reply by ghaerr** [reply]
> Can I have the codes you have modified?

Here you go. You can unzip this in elkscmd/fsck-dos then type `make`, and it should compile right up for ELKS.

[fsck-dos.zip](https://github.com/ghaerr/elks/files/14487775/fsck-dos.zip)


> I may think using far memory when I have time.

If you're serious about trying to do this, perhaps I should commit the unchanged source in to elkscmd/fsck-dos first, then the changes, so that it is easy to follow what has been changed. There are still some problems with the differences in printf format strings between systems where sizeof(int) differs, and some cleanup to be done should this be used for any production use.

The next step would be to chase down all pointers assigned to from `malloc` or `calloc`, and change them to `char ___far *` etc. Then, change to using `fmemalloc` and then convert pointers further within the program based on compilation errors. That might not be too bad a path to get it working.

The PC-98 1232k disk image could be made to work after getting 1440k working, by moving DOSBOOTBLOCKSIZE to a variable and changing a few other things as shown above.

> I have been watching your activity on Fiwix

I have been having fun with it. The Fiwix kernel is well written, and IMO could be quite suitable for a possible next-generation of machines (386+) with which to do much more with than ELKS. The author has created a very large user-space distribution that could be slimmed down for simpler uses. However, it currently lacks a TCP/IP stack (but as you may know, now supports graphics :)

Do any PC-98 machines have a 386-compatible CPU?

**Reply by tyama501** [reply]
There was PC-98 that have Celeron!
Well I only have PC-9801BX with 486 25MHz and now it is broken.

**Reply by tyama501** [reply]
Thank you for the code. Yes I think starting from unchanged codes are better for the repository.

**Reply by ghaerr** [reply]
Added in #1820. I also cleaned up the printf problems. I recommend using a typedef like `typedef char __far * lpstr` and wrapping the memory allocation functions under something like `lpstr far_calloc(uint32_t bytes)` to keep the program portable between ELKS and host to enable better debugging if needed.

**Reply by tyama501** [reply]
Thank you!
I will think about the far pointer,
but just from curiosity,
are we have chance to use "unreal mode" for 386?
(although I don't have working 386/486 real machine now)

**Reply by tyama501** [reply]
https://github.com/ghaerr/elks/blob/master/elks/arch/i86/lib/unreal.S

Even if we could enter to the unreal mode by this code, we don't have way to make 32 bits elks application with gcc-ia16, right?

**Reply by ghaerr** [reply]
> are we have chance to use "unreal mode" for 386?

The ELKS kernel already uses 386+ unreal mode when configured for XMS. However, this won't work for applications programs, and the kernel knows nothing about loading or managing 32-bit executables.

> we don't have way to make 32 bits elks application with gcc-ia16, right?

gcc-ia16 may be able to produce 32-bit ELF output, but in general there are all kinds of problems going this route, as the unreal mode segment registers are only setup once for XMS, not each time slice for when a 32-bit application runs. (In a typical 32-bit kernel, the CPU CR3 register is reset to point to an entirely new GDT each time a process context switch occurs. The GDT contains the memory mapping specific for that process).

> I will think about the far pointer

I am reading the source, it would be a good idea to study a bit. It seems that all `struct fatEntry` and `struct dosDirEntry` will likely have to be allocated in far mem. This will be a little bit expensive, due to the overhead of `fmemalloc` allocations. I think there's a 16-byte overhead for each, which will be very wasteful. It is possible it can't be made to work without huge memory use, as it looks like each FAT entry allocates a `struct fatEntry` (32 bytes+16 overhead = 48 bytes) and every directory entry (the whole tree) allocates a `struct dosDirEntry` (28+14+256+16 = 314 bytes). The 256 byte entry is for the DOS long filename, which I see now answers why so much memory is used. That could possibly be shortened, given that ELKS only handles 26 character long filenames max.

[EDIT: I just shorted the max DOS filename to 26 in the dosDirEntry struct and now it's not running out of memory... but having some other issues. Will report more likely in subsequent PR.]

**Reply by ghaerr** [reply]
Good news @tyama501, I was able to get `fsck-dos` to run on 1440k floppies without running out of RAM, and seemingly working properly. PR #1820 adds the changes to the distribution for testing.

**Reply by tyama501** [reply]
Wow, great! Thank you!

**Comment by tyama501** [comment]
Hello @ghaerr ,

I used the same corrupted image by vi for testing.
fsck-dos -pf /dev/fd1
It seems that 6 Cluster(s) lost are FIXED,
but some memory error occurred after that.

![image](https://github.com/ghaerr/elks/assets/61556504/610582bc-939b-40ae-b22e-9debf8fbda59)

No increase to the Free space,
so I assume writing back failed.

Thank you.

**Reply by tyama501** [reply]
elkscmd/fsck_dos/fat.c

writefat(int fs, struct bootblock *boot, struct fatEntry *fat, int correct_fat)

buffer = malloc(fatsz = boot->FATsecs * boot->BytesPerSec);

May be I can try allocating far memory here,
when I have more time.

**Reply by ghaerr** [reply]
Aw, shoot. Let me take another pass and see whether I can get it to work with less memory somehow. Can you post your 1440k image, or is it easy to create by just running `vi` and exiting?

**Reply by tyama501** [reply]
Here you are.
[fd1440-pc98_vi_fsck.zip](https://github.com/ghaerr/elks/files/14513157/fd1440-pc98_vi_fsck.zip)

**Reply by ghaerr** [reply]
> May be I can try allocating far memory here,

Yes - but all instances of `buffer` being passed to other functions etc have to be chased down; it is a bit complicated. I suggest using wrappers as described above for this path. We will see, perhaps it will be required to use far memory to get this to work, I had hoped I had solved it for small floppies!

**Reply by tyama501** [reply]
Now Used 1077 blocks.
It should be 1074 after fixed.

**Reply by ghaerr** [reply]
Ok @tyama501, I have hopefully fixed the out-of-memory error in #1822.

**Comment by tyama501** [comment]
Hello @ghaerr ,

I could reproduce another broken FAT using "mv".
[fd1440-pc98_mv.zip](https://github.com/ghaerr/elks/files/14541455/fd1440-pc98_mv.zip)

Well, I just moved multiple files as before...

Before mv the Used 1186.
Then I move all the files from root/ to /lib/fonts.
Now used is 1297 after all files are moved.
![df_after_mv](https://github.com/ghaerr/elks/assets/61556504/46329a92-f17e-4ee5-b72e-995949f8bcb8)

I tried fsck-dos from another elks.
fsck-dos -pf /dev/fd1
A lot of FIXED.
![fsck_dos_mv_1440](https://github.com/ghaerr/elks/assets/61556504/8a3d93e2-c421-40a1-9e57-d653cb8d2d86)

Then booted with the image again.
Great!
I don't know why the mv broke it, but
now it is Used 1187. (almost back)

![df_after_mv_fix](https://github.com/ghaerr/elks/assets/61556504/24e61e8b-ef36-46c3-81e5-92c2153555c1)

Thank you!

**Reply by toncho11** [reply]
So mv has a bug?

**Reply by ghaerr** [reply]
Hello @tyama501,

I'm a little confused by the screenshots above, in particular the `ls ../li` (note: not ../lib??), and the `mv * ../li`. There isn't a `/li` directory. Is Neko for some reason not showing the last character of a line sometimes?

> So mv has a bug?

Of course, `mv` on FAT has to do a copy, rather than a hard link, so the process SHOULD be that it first copies the files, then deletes the originals. It is appearing that it is copying the files, but NOT deleting the originals? This would explain why the free space is less after the `mv`. 

I am also guessing that `mv` is trying to `unlink` the files after they were copied, and that is halfway working: the filenames are removed from the directories, but not added back to the free list. This would produce FAT chains corresponding to unallocated files.

To test further, I would suggest trying to use `cp` followed by `rm`, using just regular filenames, not directories. Run `df` in between, and see whether each work.

We should be able to verify whether `mv` is buggy on FAT by just moving/copying/unlinking a single file, and then running `df` and `fsck-dos`.

Nonetheless, I am happy to see that our `fsck-dos` seems to be working to clean up a messed up floppy. 

> A lot of FIXED.

Yes, it appears the output is pretty verbose. I have been thinking about how to somehow condense the output so the user can see more of what is happening?

BTW, the reason that 1186 blocks used could change to 1187 (one more) after the copy/move/unlink could well be that an extra directory block was created, which might not be removed. This would be proper in the case that more than 16 files were added to a directory, (only 16 fit in a 512-byte FAT sector). Adding more would cause ELKS to add a block for the extra directory space, which would not be removed by `fsck-dos`. This could also be verified by moving/copying less than 16 files. So my guess is this 1186/1187 issue is not a problem.

Thank you!

**Reply by tyama501** [reply]
Sorry for the confusing.
I just hit the "tab" after the input of ../li.

I also made ../lib/fonts
before I move as the first picture so the directory is not created during the mv.

Thank you.
I will do things you mentioned when I have time but I think I have checked single file move does not occur this issue.

**Reply by tyama501** [reply]
Oh I haven't read correctly. Sorry.
If it more than 16 files then 1block increase.
Right?

**Reply by ghaerr** [reply]
> I think I have checked single file move does not occur this issue.

That would be interesting to know whether `mv` is buggy on a single file or not.

> If it more than 16 files then 1block increase. Right?

Yes: FAT uses 32 bytes per directory entry, so if more than 16 filenames are placed in a directory then an additional sector will have to be allocated for the next directory entries.

I've been looking further at `fsck-dos` and can clean it up a little, also reducing its memory usage. I'll post a PR for that shortly, but no new testing will be needed.

**Reply by tyama501** [reply]
To eliminate the effect of pc-98 disk driver,
this time I tested with IBM elks and qemu.
I confirmed this "mv" issue occurs on IBM elks too.

My previous post is wrong and only 1 file move can cause this problem.
misaki6.bin is about 10KB and USED increase 10 blocks after it is moved.
![mv_1file](https://github.com/ghaerr/elks/assets/61556504/04553ce0-5351-4573-9a10-02a7ea281b8a)

Also, I tried cp and rm.
cp and rm worked properly and USED decreased 10 blocks after rm.
![cp_rm_1file](https://github.com/ghaerr/elks/assets/61556504/b567e0b1-1e81-4c6a-aa0a-e3ed8a77fa7f)

BTW,
the log of fsck-dos can be seen by not applying "-p" and answering questions one by one.
I was wondering what is the difference of "-p" and "-y" for all yes,
but according to OpenBSD man it seems that "-p" is for fixing while booting without question and quit for serious errors?

**Reply by ghaerr** [reply]
Thank you for testing. I would say it confirms a definite bug with `mv` on FAT volumes that is not present with using `cp` and `rm`. I will look at the `mv` source to try to understand further.

> I was wondering what is the difference of "-p" and "-y" for all yes,
but according to OpenBSD man it seems that "-p" is for fixing while booting without question and quit for serious errors?

Yes - it seems that -p (or -fp) is useful for a quick minor check of a filesystem, with automatic fixing of small errors, but stopping on larger errors (not sure exactly what constitutes small errors, but adding lost chains back to freelist is included in small errors). This would be useful in a boot shell script, perhaps /etc/mount.cfg, where minor errors would be quickly corrected without deleting data from serious errors.

The other option -f alone will check all errors including major not handled by -p, and ask whether to fix them. If -fy is specified, then the fixes will be automatic. 

Thus, it would seem that most (automatic) fixes would be performed either using -fp or -fy options to force minor or major check.

**Reply by tyama501** [reply]
Thank you!

https://github.com/ghaerr/elks/blob/master/elkscmd/file_utils/mv.c
remove should be done with this code?

		if (unlink(srcname) < 0)
			perror(srcname);

**Reply by ghaerr** [reply]
> remove should be done with this code?

Yes. I'm looking at the code now, which does remove the "foo" when `mv foo ..` is used after the copy to ../foo, for instance. I suppose it is possible that it is the `unlink` system call on FAT that is removing a directory entry but not releasing the FAT chains? I will look into that.

**Reply by ghaerr** [reply]
@tyama501, I finally chased this down and fixed it in #1824. Perhaps after this it may be hard to create lost cluster chains in the future on FAT! :)

**Reply by tyama501** [reply]
Wow! Great, thank you!

And we were lucky that we could create lost clusters and test new fat-dos before the fix:)

**Comment by tyama501** [comment]
It seems that the boot sector off FAT12/16/32 need to have boot signature of 0xAA55 at offset 510, even the sector size is over 512 Bytes.
http://elm-chan.org/docs/fat.html#bpb

So, I think I need to add it in the boot sector of PC-98, but the boot code is already full and no space for the 2 bytes...
I need to shrink it somehow...
![image](https://github.com/ghaerr/elks/assets/61556504/33c85455-1d5e-45ed-8d9e-7adc23c4f0c8)

**Reply by ghaerr** [reply]
I can’t read that link, but if your FAT floppies work with PC-98 DOS, I would not change the boot sector, especially just for fsck-dos. Just modify fsck-dos instead to make it work with your unmodified floppy.

**Reply by tyama501** [reply]
Hello @ghaerr ,

MS-DOS for PC-98 cannot read elks disk. (Unclear. May be boot signature is not the only reason.)
FreeDOS(98) can read elks disk.

I tried code you have posted on https://github.com/ghaerr/elks/pull/1822#issuecomment-1983915319

https://github.com/tyama501/elks/commit/7fbcf00a5faff6ece1844af23eeb84256d3609c4

It worked well to fix lost chain in the fd1232.img (not shown but disk in fd1 was USED 1109 91% before fix)
![fsck1232](https://github.com/ghaerr/elks/assets/61556504/a5e3546d-380c-4cfa-b7c2-ce7a3954d311)

Thank you!

**Reply by tyama501** [reply]
(Unclear. May be boot signature is not the only reason.)

I remember I got a report on twitter that it needs FAT media Byte in FAT[0] LSB.

**Reply by ghaerr** [reply]
If you can post a hex dump of both ELKS and PC-98 boot sector we can analyze to check differences.

**Reply by ghaerr** [reply]
Hello @tyama501,

There are several issues you've brought up, I'll comment on each:

> I tried code you have posted on https://github.com/ghaerr/elks/pull/1822#issuecomment-1983915319
> It worked well to fix lost chain in the fd1232.img (not shown but disk in fd1 was USED 1109 91% before fix)

Go ahead and post a PR with the fsck-dos fix, except also fix my typo in the line:
```
perror("Invalid DOS format, bad bytes per sector");
```
to
```
pfatal("Invalid DOS format, bad bytes per sector");
```

> It seems that the boot sector off FAT12/16/32 need to have boot signature of 0xAA55 at offset 510, even the sector size is over 512 Bytes.
> So, I think I need to add it in the boot sector of PC-98, but the boot code is already full and no space for the 2 bytes...
I need to shrink it somehow...

Are you saying the PC-98 boot code for the 1232k floppy (with 1024-byte sectors) is full? There should be plenty of room left. I am aware that the PC-98 1440k floppy with 512-byte sectors is very full, but it does have the 0xAA55 at offset 510, correct?

> It seems that the boot sector off FAT12/16/32 need to have boot signature of 0xAA55 at offset 510, even the sector size is over 512 Bytes.
> I remember I got a report on twitter that it needs FAT media Byte in FAT[0] LSB.

If you can post a hex dump of a working DOS 1232k boot sector (1024 bytes) as well as that of the ELKS 1232k boot sector, we can compare to determine whether a modification might be made to add 0xAA55 at offset 510, as well as what the FAT media byte is between the two floppies.

EDIT:

A quick look at bootblocks/boot_sect.S seems to show that the opposite of what we are thinking is happening - that 0xAA55 is at offset 510, but NOT at offset 1022 for 1232k floppies. Adding another `.org SECTOR_SIZE-2` ifdefed for CONFIG_IMG_1232 might solve the problem.

Thank you!

**Reply by tyama501** [reply]
Thank you @ghaerr 

I will make PR.

By looking with binary editor the elks parameters and boot signature are in around 1022.

1232 bootsector for the first 512 Bytes are full because it is different code from 1440.
(One idea is shrink ELKS! to EL!)

I am suspecting the PC-98 DOS issue is related to the value in FAT0 Allocate table.
I will keep looking it.

**Reply by ghaerr** [reply]
> By looking with binary editor the elks parameters and boot signature are in around 1022.

I see now, you're right. 

> 1232 bootsector for the first 512 Bytes are full because it is different code from 1440.
(One idea is shrink ELKS! to EL!)

Good idea, go ahead and shorten the message and add the boot signature at 510. If a shorter boot message works to make PC-98 ELKS FAT disks compatible with DOS, that would be worth it.

> I am suspecting the PC-98 DOS issue is related to the value in FAT0 Allocate table.

Oh I see, we're not talking about the BPB media byte, but the media byte in the first position of the FAT table. You should be able to use hex editor to see if that might be the issue, and if so, we can change our tools to update that to the proper value. Also, if media byte is the case, then the fsck-dos code might have to be changed in your PR:
```
if ((boot->BytesPerSec == 512) && (block[510] != 0x55 || block[511] != 0xaa)) {
```
to
```
if (block[510] != 0x55 || block[511] != 0xaa) {
```
(that is, not special cased and not changed by your PR). So perhaps better to not post PR until we know whether boot signature needs to be at 510, or its just a FAT[0] media byte problem.

**Reply by tyama501** [reply]
Hello @ghaerr ,

I could read elks disk fd1232 from DOS 6.2 by just changing the single Byte in FAT[0] from 0xF0 to 0xFE.

![fd1232_media_FE](https://github.com/ghaerr/elks/assets/61556504/e808392d-bd63-4ddd-ad38-e6f66fe67287)

**Reply by ghaerr** [reply]
> I could read elks disk fd1232 from DOS 6.2 by just changing the single Byte in FAT[0] from 0xF0 to 0xFE.

Very interesting and good news! Let me try to find some documentation the required/allowed values and meanings for the FAT[0] byte. I'm pretty sure all the FAT disks ELKS produces are created using the external `mtools-4.0.24` package. We may require a custom version of it in order to change this byte for PC-98 1232 disks.

Do you have any documentation that shows the requirement for 0xFE as the FAT[0] media byte?

**Reply by ghaerr** [reply]
Also can you lookup the value of the BPB media byte at offset 21 (decimal) of the boot sector? I am interested to know if this is also 0xFE, or something else.

**Reply by ghaerr** [reply]
@tyama501:

It looks like we may be able to set the media byte (in both the FAT[0] position, apparently called FAT ID as well as in the BPB, called media descriptor byte) to 0xFE by the `mformat` command used by ELKS when creating the PC-98 1232k floppy. In order to do this, edit the file images/Make.defs and try changing from:
```
ifdef CONFIG_IMG_FD1232
FAT_MKFSOPTS = -s 8 -h 2 -t 77 -M 1024 -d 2 -L 2 -r 6 -k -N 0 -c 1
```
to (adding -m 0xfe to the end):
```
ifdef CONFIG_IMG_FD1232
FAT_MKFSOPTS = -s 8 -h 2 -t 77 -M 1024 -d 2 -L 2 -r 6 -k -N 0 -c 1 -m 0xfe
```
I don't think ELKS looks at the media descriptor byte nor FAT ID, so this should be OK. If this works, please include this in your PR, along with what you learn about the need for checking for the boot signature at offset 510.


Thank you!

**Reply by tyama501** [reply]
Thank you!
https://github.com/ghaerr/elks/pull/1851#issuecomment-2041179114
